common/src/main/java/org/conscrypt/NativeSsl.java

```diff
@@ -375,6 +357,44 @@ void initialize(String hostname, OpenSSLKey channelIdPrivateKey) throws IOExcept
         setTlsChannelId(channelIdPrivateKey);
     }
 
+    void configureServerCertificate() throws IOException {
+        verifyWithSniMatchers(getRequestedServerName());
+        if (isClient()) {
+            return;
+        }
+        X509KeyManager keyManager = parameters.getX509KeyManager();
+        if (keyManager != null) {
+            for (String keyType : getCipherKeyTypes()) {
+                try {
+                    setCertificate(aliasChooser.chooseServerAlias(keyManager, keyType));
+                } catch (CertificateEncodingException e) {
+                    throw new IOException(e);
+                }
+            }
+        }
+    }
+
+    private void verifyWithSniMatchers(String serverName) {
+        if (!AddressUtils.isValidSniHostname(serverName)) {
+            return;
+        }
+
+        if (Platform.allSniMatchersFail(parameters, serverName)) {
```

Comment:
The negative name of this function (`allSniMatchersFail`) is awkward, though I understand you want to capture the fact that an empty list matches everything.  I think something like `serverNamePermitted` would still capture the semantics while being easier to read.